src/triage/component/postmodeling/utils/predictions_selected_model_groups.py

```diff
@@ -0,0 +1,175 @@
+        
+    
+def generate_predictions(db_engine, model_groups, project_path, experiment_hashes=None, train_end_times_range=None, rank_order='worst', replace=False):
+    """ Generate predictions and write to DB for a set of model_groups in an experiment
+        Args:
+            db_conn: Sqlalchemy engine
+            model_groups (list): The list of model group ids we are interested in (ideally, chosen through audition)
+            project_path (str): Path where the created matrices and trained model objects are stored for the experiment
+            experiment_hashes (List[str]): Optional. hash of the experiment (currently handling only one experiment)
+            range_train_end_times (Dict): Optional. If given, only the models with train_end_times that fall in the range are scored.
+                                        A dictionary with two possible keys 'range_start_date' and 'range_end_date'. Either or both could be set
+            rank_order (str) : How to deal with ties in the scores. 
+            replace (bool) : Whether to overwrite the preditctions for a model_id, if already found in the DB
+
+        Returns: None
+            This directly writes to the test_results.predictions table
+    """
+
+    model_matrix_info = _fetch_relevant_model_matrix_info(
+        db_engine=db_engine,
+        model_groups=model_groups,
+        experiment_hashes=experiment_hashes
+    )
+
+    if len(model_matrix_info)==0:
+        raise ValueError('No models were found for the given experiment and model group(s)')
+    
+    # All the model groups specified in the  
+    not_fetched_model_grps = [x for x in model_groups if not x in model_matrix_info['model_group_id'].unique()]
+    
+    if len(not_fetched_model_grps) > 0:
+        raise ValueError('No models were found for the model group(s) {}. All specified model groups should be present'.format(not_fetched_model_grps))
+    
+    logging.info('Found {} model ids'.format(len(model_matrix_info)))
+
+    # If we are only generating predictions for a specific time range
+    if train_end_times_range is not None: 
+        if 'range_start_date' in train_end_times_range:
+            range_st = train_end_times_range['range_start_date']
+            msk = (model_matrix_info['train_end_time'] >= range_st)
+            logging.info('Filtering out models with a train_end_time before {}'.format(range_st))
+
+            model_matrix_info = model_matrix_info[msk]
+
+        if 'range_end_date' in train_end_times_range:       
+            range_en = train_end_times_range['range_end_date']
+            msk = (model_matrix_info['train_end_time'] <= range_en)
+            logging.info('Filtering out models with a train_end_time after {}'.format(range_en))
+
+            model_matrix_info = model_matrix_info[msk]
+
+        if len(model_matrix_info) == 0:
+            raise ValueError('No models were found for the given time range')
```

Comment:
Maybe consolidate with the check above and only do this once after doing all the filtering?